cmd/verify_metadata.go

```diff
@@ -144,41 +150,137 @@ func verifyMetadata(cmd *cobra.Command, args []string) error {
+    if err != nil {
+        return fmt.Errorf("failed to read artifact file: %v", err)
+    }
+    var jsonData *amalgamView
+    if err = json.Unmarshal(bytes, &jsonData); err != nil {
+        return fmt.Errorf("failed to unmarshal JSON data: %v", err)
+    }
+
+    views := []struct {
+        Name             string
+        Info             string
+        SignatureDetails SigDetails
+    }{
+        {"BuildInfo", jsonData.BuildInfo, jsonData.BuildInfoSignature},
+        {"HealthInfo", jsonData.HealthInfo, jsonData.HealthInfoSignature},
+        {"VexInfo", jsonData.VexInfo, jsonData.VexInfoSignature},
+    }
+
+    for _, v := range views {
+        if v.Info == "" && v.Name == "HealthInfo" {
+            continue
+        }
+        cert, err := parseCertificate([]byte(v.SignatureDetails.CertInfo.Cert))
+        if err != nil {
+            return err
+        }
+
+        // Verify certificates.
+        if !disableCertificateVerification {
+            // Download root certificate.
+            certPath := filepath.Join(destDir, "ca.crt")
+            if err := downloadRootCert(certPath); err == nil {
+                cmd.Printf("File downloaded at %s\n", certPath)
+            } else {
+                return err
+            }
+
+            if ok, err := verifyCertificate([]byte(v.SignatureDetails.CertInfo.CertChain), certPath, cert); ok {
+                cmd.Printf("%s Certificates Verified successfully!\n", v.Name)
+            } else {
+                cmd.Printf("Unsuccessful Certificate Verification of %s\n", v.Name)
+                if err != nil {
+                    return err
+                }
+            }
+        }
+
+        // Verify data integrity.
+        ok, err := verifyDigest([]byte(v.Info), v.SignatureDetails.Digest[0].Digest)
+        if !ok {
+            if err != nil {
+                return err
+            }
+            return fmt.Errorf("incorrect Digest of %v", v.Name)
+        }
+
+        // Verify authenticity.
+        sig, err := hex.DecodeString(v.SignatureDetails.Signature[0].Signature)
+        if err != nil {
+            return fmt.Errorf("failed to decode signature of %v: %v", v.Name, err)
+        }
+        dig, err := hex.DecodeString(v.SignatureDetails.Digest[0].Digest)
+        if err != nil {
+            return fmt.Errorf("failed to decode digest of %v: %v", v.Name, err)
+        }
+        ok, err = verifySignatures([]byte(sig), []byte(dig), cert)
+        if ok {
+            cmd.Printf("%s Metadata Signature Verified successfully!\n", v.Name)
+        } else {
+            cmd.Printf("%s Unsuccessful Metadata Signature Verification\n", v.Name)
+            if err != nil {
+                return err
+            }
+        }
+
+    }
+    return nil
+}
+
+func verifyNONPremiumMetadata(cmd *cobra.Command, serviceAccountKeyFilePath, destDir, language, packageID, version, metadata_type string, disableCertificateVerification, disableDeletes bool) error {
```

Comment:
we can create struct for input. Also can you check if verifyNONPremiumMetadata and verifyPremiumMetadata shares any common code that can be abstract as a separate function.

Reply:
done.